File: leetcode_java/src/main/java/com/programapprentice/app/FindMinimumInRotatedSortedArray.java
package com.programapprentice.app;

/**
 * User: ruijiang
 * Date: 10/21/14
 * Time: 6:46 PM
 */
/**
 * Suppose a sorted array is rotated at some pivot unknown to you beforehand.
 * (i.e., 0 1 2 4 5 6 7 might become 4 5 6 7 0 1 2).
 * Find the minimum element.
 * You may assume no duplicate exists in the array.
 * */
public class FindMinimumInRotatedSortedArray {
    public int findMin(int[] num) {
        if(num == null || num.length == 0) {
            return 0;
        }
        if(num.length == 1) {
            return num[0];
        }
        int low = 0;
        int high = num.length-1;
        int mid = 0;
        while(low <= high) {
            mid = (high + low) / 2;
            // if mid is min, return it. Otherwise it needs to move high and low
            if(num[(mid - 1 + num.length) % num.length] > num[mid] && num[(mid + 1) % num.length] > num[mid]) {
                return num[mid];
            }
            if(num[mid] > num[high]) {
                low = (mid + 1) % num.length;
            } else {
                high = (mid - 1 + num.length) % num.length;
            }
        }
        return num[low];
    }
}

File: leetcode_cpp/src/find_minimum_in_rotated_sorted_array.cc
/*
 * find_minimum_in_rotated_sorted_array.cc
 *
 *  Created on: Nov 22, 2014
 *      Author: ruijiang
 */

#include "find_minimum_in_rotated_sorted_array.h"


int Solution::findMin(vector<int> &num){
    int length = num.size();
    if(0 == length) {
        return num[0];
    }
    int low = 0;
    int high = length - 1;
    int mid = 0;
    int midlow = 0;
    int midhigh = 0;
    while(low < high) {
        mid = (low + high) / 2;
        midlow = (length + mid - 1) % length;
        midhigh = (mid + 1) % length;
        if(num[midlow] > num[mid] && num[midhigh] > num[mid]) {
            return num[mid];
        }
        if(num[high] > num[low]) {
            if(num[high] > num[mid]) {
                high = midlow;
            } else {
                low = midhigh;
            }
        } else {
            if(num[high] > num[mid]) {
                high = midlow;
            } else {
                low = midhigh;
            }
        }
    }
    return num[low];
}


// http://www.cnblogs.com/forcheryl/p/4032278.html
//int Solution::findMin(vector<int> &num) {
//    int start=0,end=num.size()-1;
//
//    while (start<end) {
//        if (num[start]<num[end])
//            return num[start];
//
//        int mid = (start+end)/2;
//
//        if (num[mid]>=num[start]) {
//            start = mid+1;
//        } else {
//            end = mid;
//        }
//    }
//
//    return num[start];
//}
//
